db/python/layers/ourdna/dashboard.py

```diff
@@ -0,0 +1,371 @@
+# pylint: disable=too-many-locals
+import asyncio
+from collections import defaultdict
+from datetime import datetime
+from math import ceil
+from typing import Any
+
+from db.python.connect import Connection
+from db.python.layers.base import BaseLayer
+from db.python.layers.participant import ParticipantLayer
+from db.python.layers.sample import SampleLayer
+from db.python.tables.sample import SampleFilter
+from db.python.utils import GenericFilter
+from models.models import OurDNADashboard, OurDNALostSample, ProjectId, Sample
+from models.models.participant import ParticipantInternal
+
+
+class OurDnaDashboardLayer(BaseLayer):
+    """Layer for analysis logic"""
+
+    def __init__(self, connection: Connection):
+        super().__init__(connection)
+
+        self.sample_layer = SampleLayer(connection)
+        self.participant_layer = ParticipantLayer(connection)
+
+    @staticmethod
+    def get_meta_property(sample: Sample, property_name: str) -> Any:
+        """
+        Get a property from the meta field of a sample, accounting for hyphenated property names
+        or underscores in the property name
+        """
+        return sample.meta.get(property_name) or sample.meta.get(
+            property_name.replace('-', '_')
+        )
+
+    def get_collection_to_process_end_time(self, sample: Sample) -> int | None:
+        """
+        I want to know how long it took between blood collection and sample processing - SAMPLE TABLE
+        @fields: collection-time, process-end-time
+        """
+        _collection_time = self.get_meta_property(
```

Comment:
I think this logic would be become simpler if there's a pre-processing step that pulls out these fields as pre-parsed dates, and then these values can be methods on that class. Pulling various tags from meta continuously, and parsing the dates each time feels a little tedious.

Maybe run a pre-processing step to get:

```python
class SampleProcessMeta:
    collection_time: datetime | None
    process_site: str
    process_start_time: datetime | None
    process_end_time: datetime | None

    @staticmethod
    def try_parse_datestr(datestr):
        try:
            return datetime.strptime(
                datestr, '%Y-%m-%d %H:%M:%S'
            )
        except:
            return None

    @staticmethod
    def from_sample(sample: Sample):
        collection_time_str = get_meta_property(sample=sample, property_name='collection-time')
        try:
        
        # ...
        return SampleProcessMeta(...)

    def get_collection_to_process_end_time_seconds(self) -> int:
        if not self.collection_time or not self.process_end_time:
            return None
        return int(self.process_end_time - self.collection_time)
```

What do you think?